context.go

```diff
@@ -731,18 +735,33 @@ func (c *Context) WalkBlueprintsFiles(rootDir string, filePaths []string,
 	// Darwin has a default limit of 256 open files
 	maxActiveCount := 200
 
+	// count the number of pending calls to visitor()
+	visitorWaitGroup := sync.WaitGroup{}
+
 	startParseBlueprintsFile := func(blueprint stringAndScope) {
-		if blueprintsSet[blueprint.string] {
+		if blueprintsSet[blueprint.fileName] {
 			return
 		}
-		blueprintsSet[blueprint.string] = true
+		blueprintsSet[blueprint.fileName] = true
 		activeCount++
-		deps = append(deps, blueprint.string)
+		deps = append(deps, blueprint.fileName)
+		visitorWaitGroup.Add(1)
 		go func() {
-			c.parseOneAsync(blueprint.string, blueprint.Scope, rootDir,
-				errsCh, fileCh, blueprintsCh, depsCh)
+			file := c.parseOneAsync(blueprint.fileName, blueprint.Scope, rootDir,
+				errsCh, blueprintsCh, depsCh, blueprint.doneVisiting)
+			doneParsingCh <- blueprint
 
-			doneCh <- blueprint
+			if blueprint.readyToVisit != nil {
+				// wait for visitor() of parent to complete
+				<-blueprint.readyToVisit
+			}
+
+			// process this file
+			visitor(file)
+			if blueprint.doneVisiting != nil {
```

Comment:
IIUC doneVisiting is the same channel as readyToVisit in the descendants?  If so, that should be commented somewhere.

Coudl the blueprint object have a pointer to the parent blueprint object, and the readyToVisit check would become <-blueprint.parent.doneVisiting?

Reply:
I've now implemented the latter